packages/sources/universal-v3/src/utils/LambdaController.ts

```diff
@@ -0,0 +1,348 @@
+      MemorySize: this.config.LAMBDA_MEMORY_SIZE_MB,
+      Environment: {
+        Variables: {
+          LOG_LEVEL: this.config.LAMBDA_LOG_LEVEL,
+        },
+      },
+      // VpcConfig: // This setting allows for putting the Lambda behind a VPC
+    }
+
+    return new Promise((resolve, reject) => {
+      this.lambda.createFunction(params, (err, data) => {
+        if (err) {
+          this.handleLambdaErrorResponse(
+            err,
+            retryCount,
+            FunctionName,
+            sandboxRequestData,
+            resolve,
+            reject,
+          )
+          return
+        }
+
+        this.logger.debug(`Lambda creation response: ${JSON.stringify(data)}`)
+
+        // Disabling no-non-null-assert since executeRequestInLambda already ensures initalization
+        /* eslint-disable @typescript-eslint/no-non-null-assertion */
+        this.deployedLambdaFunctions!.timeLastUsed[FunctionName] = Date.now()
+        this.deployedLambdaFunctions!.count++
+        /* eslint-enable @typescript-eslint/no-non-null-assertion */
+
+        setTimeout(
+          () => resolve(this.executeRequestInLambda(FunctionName, sandboxRequestData, retryCount)),
+          this.config.LAMBDA_INIT_TIME_MS,
+        )
+      })
+    })
+  }
+
+  // This recursive function returns all deployed Lambda functions as a Record from the name to
+  // the time the function was last used.  The TimeLastUsed is initally set to zero.
+  // Documentation on lambda.listFunctions: https://docs.aws.amazon.com/lambda/latest/dg/API_ListFunctions.html
+  private getDeployedLambdaFunctions = async (
+    Marker?: string,
+  ): Promise<DeployedLambdaFunctions> => {
+    return new Promise((resolve) => {
+      this.lambda.listFunctions({ Marker }, async (err, data) => {
+        if (err) {
+          this.logger.error(`Unexpected Lambda listFunctions error: ${err}`)
+          throw Error(`Error fetching list of deployed Lambda functions: ${err}`)
+        }
+
+        const functions = data.Functions
+
+        if (!functions) {
+          resolve({ count: 0, timeLastUsed: {} })
+          return
+        }
+
+        let deployedLambdaFunctions: DeployedLambdaFunctions = { count: 0, timeLastUsed: {} }
+
+        // Each call to getDeployedLambdaFunctions fetches up to 50 Lambda functions at a time.
+        // If more than 50 Lambda functions exist, the next page must be fetched using the NextMarker.
+        if (data.NextMarker) {
+          deployedLambdaFunctions = await this.getDeployedLambdaFunctions(data.NextMarker)
+        }
+
+        functions.forEach(({ FunctionName }) => {
+          if (FunctionName) deployedLambdaFunctions.timeLastUsed[FunctionName] = 0
+        })
+
+        deployedLambdaFunctions.count = Object.keys(deployedLambdaFunctions.timeLastUsed).length
+
+        resolve(deployedLambdaFunctions)
+      })
+    })
+  }
+
+  private getLeastRecentlyUsedFunction = async (): Promise<string> => {
```

Comment:
This is true, however if you implement your suggestion, you are incurring more compute cost each time the function is called as the array needs to be re-ordered.

I instead chose the simple approach and decided to incur the cost on getLeastRecentlyUsedFunction as this will be called less frequently than invokeLambdaFunction.

Reply:
You don't need to reorder an array, implementing a simple double linked list makes shifting a known node to the head or tail a constant time operation (there are more efficient methods still, that's just an easy one). It's true that it adds a few OPs more on each invocation, that's why I was asking about the number of items that will be relevant here, but yeah if it's low 4 digits and this is called infrequently I agree it's not worth the extra complexity atm